libvast/builtins/operators/put.cpp

```diff
@@ -0,0 +1,228 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/arrow_table_slice.hpp>
+#include <vast/concept/parseable/to.hpp>
+#include <vast/concept/parseable/vast/data.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/detail/narrow.hpp>
+#include <vast/error.hpp>
+#include <vast/legacy_pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/table_slice_builder.hpp>
+#include <vast/type.hpp>
+
+#include <arrow/array.h>
+#include <fmt/format.h>
+
+namespace vast::plugins::put {
+
+namespace {
+
+/// The parsed configuration.
+struct configuration {
+  std::vector<std::tuple<std::string, data>> extractor_to_value = {};
+};
+
+/// The confguration bound to a specific schema.
+struct bound_configuration {
+  /// Bind a *configuration* to a given schema.
+  /// @param schema The schema to bind to.
+  /// @param config The parsed configuration.
+  static caf::expected<bound_configuration>
+  make(const type& schema, const configuration& config) {
+    auto result = bound_configuration{};
+    const auto& schema_rt = caf::get<record_type>(schema);
+    auto extensions = std::vector<std::tuple<std::string, data, type>>{};
+    for (const auto& [extractor, value] : config.extractor_to_value) {
+      bool found = false;
+      // If the extractor resolves, we replace all matched fields.
+      for (const auto& index :
+           schema_rt.resolve_key_suffix(extractor, schema.name())) {
+        found = true;
+        result.transformations.push_back({index, make_replace(value)});
+      }
+      // If the extractor did not resolve, we instead add one new field at the
+      // end.
+      if (not found) {
+        auto inferred_type = type::infer(value);
+        if (not inferred_type)
+          return caf::make_error(ec::logic_error,
+                                 fmt::format("failed to infer type from '{}'",
+                                             value));
+        extensions.emplace_back(extractor, value, std::move(inferred_type));
+      }
+    }
+    if (not extensions.empty())
+      result.transformations.push_back(
+        {{schema_rt.num_fields() - 1}, make_extend(std::move(extensions))});
```

Comment:
Can we have `schema_rt.num_fields() == 0`?

Reply:
Record types by definition must have at least one field currently.